src/cata_ui.cpp

```diff
@@ -0,0 +1,1038 @@
+                anchored_y = rect.y;
+                return;
+            case center_left:
+                anchored_x = rect.x;
+                anchored_y = ( p_rect.size_y / 2 ) - ( rect.size_y / 2 ) + rect.y;
+                return;
+            case center_center:
+                anchored_x = ( p_rect.size_x / 2 ) - ( rect.size_x / 2 ) + rect.x;
+                anchored_y = ( p_rect.size_y / 2 ) - ( rect.size_y / 2 ) + rect.y;
+                return;
+            case center_right:
+                anchored_x = p_rect.size_x - rect.size_x + rect.x;
+                anchored_y = ( p_rect.size_y / 2 ) - ( rect.size_y / 2 ) + rect.y;
+                return;
+            case bottom_left:
+                anchored_x = rect.x;
+                anchored_y = p_rect.size_y - rect.size_y + rect.y;
+                return;
+            case bottom_center:
+                anchored_x = ( p_rect.size_x / 2 ) - ( rect.size_x / 2 ) + rect.x;
+                anchored_y = p_rect.size_y - rect.size_y + rect.y;
+                return;
+            case bottom_right:
+                anchored_x = p_rect.size_x - rect.size_x + rect.x;
+                anchored_y = p_rect.size_y - rect.size_y + rect.y;
+                return;
+        }
+    }
+    anchored_x = rect.x;
+    anchored_y = rect.y;
+}
+
+unsigned int ui_element::get_ax() const
+{
+    return anchored_x;
+}
+
+unsigned int ui_element::get_ay() const
+{
+    return anchored_y;
+}
+
+void ui_element::set_parent( const ui_window *parent )
+{
+    this->parent = parent;
+    calc_anchored_values();
+}
+
+const ui_rect &ui_element::get_rect() const
+{
+    return rect;
+}
+
+WINDOW *ui_element::get_win() const
+{
+    if( parent != nullptr ) {
+        return parent->get_win();
+    }
+    return nullptr;
+}
+
+ui_label::ui_label( std::string text ,int x, int y, ui_anchor anchor ) : ui_element( utf8_width( text.c_str() ), 1, x, y, anchor ), text( text )
+{
+}
+
+void ui_label::draw()
+{
+    auto win = get_win();
+    if( win == nullptr ) {
+        return;
+    }
+
+    mvwprintz( win, get_ay(), get_ax(), text_color, text.c_str() );
+}
+
+void ui_label::set_text( std::string new_text )
+{
+    text = new_text;
+    set_rect( ui_rect( new_text.size(), get_rect().size_y, get_rect().x, get_rect().y ) );
+}
+
+bordered_window::bordered_window( size_t size_x, size_t size_y, int x, int y, ui_anchor anchor ) : ui_window( size_x, size_y, x, y, anchor )
+{
+}
+
+void bordered_window::local_draw()
+{
+    auto win = get_win(); // never null
+
+    wattron( win, border_color );
+    wborder( win, LINE_XOXO, LINE_XOXO, LINE_OXOX, LINE_OXOX,
+            LINE_OXXO, LINE_OOXX, LINE_XXOO, LINE_XOOX );
+    wattroff( win, border_color );
+}
+
+health_bar::health_bar( size_t size_x, int x, int y, ui_anchor anchor ) : ui_element( size_x, 1, x, y, anchor ),
+                       max_health( size_x * points_per_char ), current_health( max_health )
+{
+    for( unsigned int i = 0; i < get_rect().size_x; i++ ) {
```

Comment:
There is a string constructor you can use: `std::string(count, character)`, it does exactly what you do with the loop here.

Might also be useful at line 322.